Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -166,3 +167,22 @@ extension SOCKSClientHandler {
     }
     
 }
+
+extension SOCKSClientHandler: RemovableChannelHandler {
+    
+    public func removeHandler(context: ChannelHandlerContext, removalToken: ChannelHandlerContext.RemovalToken) {
+        guard self.state.proxyEstablished else {
+            preconditionFailure("The SOCKSClientHandler can only be removed once a connection has been established")
+        }
```

Comment:
We could work around this precondition by adding this to the state machine. Thus we would be waiting for the SOCKS handshake to succeed or fail and only trigger the leave then. @Davidde94 @Lukasa is this something we want to add?

Reply:
Adding what to the state machine? A new action "remove handler"?

Reply:
Yeah, I think that works.

Reply:
Discussed offline - I think that the state machine should be completely decoupled from channel handler logic. The proposed solution was to store an optional removal token in the handler, and deal with removal when the state machine emits `proxy established`.